File: experiments/java-sdk/examples/com/datadog/llmobs/experiments/examples/DatasetOperations.java
package com.datadog.llmobs.experiments.examples;

import com.datadog.llmobs.experiments.Dataset;
import com.datadog.llmobs.experiments.DatasetRecord;
import com.datadog.llmobs.experiments.ExperimentsClient;

import java.util.Map;

/**
 * Standalone dataset example: create a dataset, push a few records, then pull the
 * same dataset back to verify the round-trip. No experiment is run here — this is
 * intentionally just the data layer, useful for getting a feel for {@link Dataset}
 * in isolation.
 *
 * <p>Run with: {@code ./gradlew runDatasetOperations}
 */
public final class DatasetOperations {

    public static void main(String[] args) {
        String apiKey = requireEnv("DD_API_KEY");
        String appKey = requireEnv("DD_APPLICATION_KEY");
        String site = System.getenv().getOrDefault("DD_SITE", "datadoghq.com");

        ExperimentsClient client = ExperimentsClient.builder()
            .apiKey(apiKey)
            .applicationKey(appKey)
            .site(site)
            .projectName("java-sdk-bootstrap")
            .build();

        String datasetName = "dataset-ops-demo-" + System.currentTimeMillis();

        // -- Create + push -----------------------------------------------------
        System.out.println("=== Creating dataset and pushing 3 records ===");
        Dataset created = client.createDataset(datasetName, "Demo of create/push/pull")
            .addRecord(
                Map.of("question", "What is 2+2?"),
                "4",
                Map.of("source", "arithmetic", "difficulty", "easy")
            )
            .addRecord(
                Map.of("question", "What is the capital of France?"),
                "Paris",
                Map.of("source", "geography", "difficulty", "medium")
            )
            .addRecord(
                Map.of("question", "Define photosynthesis."),
                "The process by which plants convert light into energy.",
                Map.of("source", "biology", "difficulty", "hard")
            );

        created.push();   // explicit — no Experiment.run() needed
        System.out.println("Created dataset id : " + created.id());
        System.out.println("Pushed records     : " + created.records().size());
        System.out.println();

        // -- Pull back ----------------------------------------------------------
        System.out.println("=== Pulling the same dataset back from Datadog ===");
        Dataset pulled = client.pullDataset(datasetName);
        System.out.println("Pulled dataset id  : " + pulled.id());
        System.out.println("Pulled records     : " + pulled.records().size());
        int idx = 0;
        for (DatasetRecord r : pulled.records()) {
            System.out.println("  [" + idx + "] input=" + r.input()
                + "  expected=" + r.expectedOutput()
                + "  metadata=" + r.metadata());
            idx++;
        }

        // -- Verify -------------------------------------------------------------
        System.out.println();
        if (created.id().equals(pulled.id()) && created.records().size() == pulled.records().size()) {
            System.out.println("Round-trip OK: created id matches pulled id, record counts match.");
        } else {
            System.out.println("Mismatch: created.id=" + created.id() + " pulled.id=" + pulled.id()
                + " created.count=" + created.records().size()
                + " pulled.count=" + pulled.records().size());
        }
    }

    private static String requireEnv(String name) {
        String v = System.getenv(name);
        if (v == null || v.isBlank()) {
            System.err.println("Missing required env var: " + name);
            System.exit(1);
        }
        return v;
    }
}

File: experiments/node-sdk/examples/datasetOperations.ts
/**
 * DatasetOperations — dataset create/push/pull round-trip over three records.
 * Creates and pushes a dataset, then pulls the same dataset back by name and
 * verifies the round-trip. Writes to the `node-sdk-bootstrap` project.
 *
 *   DD_API_KEY=... DD_APPLICATION_KEY=... npm run runDatasetOperations
 */
import { ExperimentsClient } from "../src/index.js";

function requireEnv(name: string): string {
  const v = process.env[name];
  if (!v || v.trim() === "") {
    console.error(`Missing required env var: ${name}`);
    process.exit(1);
  }
  return v;
}

async function main(): Promise<void> {
  const client = new ExperimentsClient({
    apiKey: requireEnv("DD_API_KEY"),
    applicationKey: requireEnv("DD_APPLICATION_KEY"),
    site: process.env.DD_SITE ?? "datadoghq.com",
    projectName: "node-sdk-bootstrap",
  });

  const datasetName = `dataset-ops-demo-${Date.now()}`;

  console.log("=== Creating dataset and pushing 3 records ===");
  const dataset = client
    .createDataset(datasetName, "Demo of create/push/pull")
    .addRecord({ question: "What is 2+2?" }, "4", { source: "arithmetic", difficulty: "easy" })
    .addRecord({ question: "What is the capital of France?" }, "Paris", {
      source: "geography",
      difficulty: "medium",
    })
    .addRecord(
      { question: "Define photosynthesis." },
      "The process by which plants convert light into energy.",
      { source: "biology", difficulty: "hard" },
    );
  await dataset.push();
  console.log(`Created dataset id : ${dataset.id()}`);
  console.log(`Dataset URL        : ${dataset.url()}`);
  console.log(`Pushed records     : ${dataset.records().length}`);

  console.log("=== Pulling the same dataset back from Datadog ===");
  // Wait (exponential backoff, up to 30s) until all pushed records are readable —
  // LLM Obs reads are eventually consistent right after a write.
  const pulled = await client.pullDataset(datasetName, {
    expectedRecordCount: dataset.records().length,
  });
  console.log(`Pulled dataset id  : ${pulled.id()}`);
  console.log(`Pulled records     : ${pulled.records().length}`);
  pulled.records().forEach((rec, i) => {
    console.log(
      `  [${i}] input=${JSON.stringify(rec.input)}  expected=${JSON.stringify(rec.expectedOutput)}  metadata=${JSON.stringify(rec.metadata)}`,
    );
  });

  if (dataset.id() === pulled.id() && dataset.records().length === pulled.records().length) {
    console.log("Round-trip OK: created id matches pulled id, record counts match.");
  } else {
    console.log(
      `Mismatch: created.id=${dataset.id()} pulled.id=${pulled.id()} created.count=${dataset.records().length} pulled.count=${pulled.records().length}`,
    );
  }
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});
